game-core/src/main/java/games/strategy/triplea/delegate/battle/FireAa.java

```diff
@@ -92,84 +89,77 @@
-          CollectionUtils.getMatches(firingUnits, Matches.unitIsAaOfTypeAa(aaType));
-      final List<Collection<Unit>> firingGroups = MustFightBattle.newFiringUnitGroups(aaTypeUnits);
-      for (final Collection<Unit> firingGroup : firingGroups) {
-        final Set<UnitType> validTargetTypes =
-            UnitAttachment.get(firingGroup.iterator().next().getType())
-                .getTargetsAa(bridge.getData());
-        final Set<UnitType> airborneTypesTargeted =
-            defending
-                ? TechAbilityAttachment.getAirborneTargettedByAa(hitPlayer, bridge.getData())
-                    .get(aaType)
-                : new HashSet<>();
-        final Collection<Unit> validTargets =
-            CollectionUtils.getMatches(
-                attackableUnits,
-                Matches.unitIsOfTypes(validTargetTypes)
-                    .or(
-                        Matches.unitIsAirborne()
-                            .and(Matches.unitIsOfTypes(airborneTypesTargeted))));
-        final IExecutable rollDice =
-            new IExecutable() {
-              private static final long serialVersionUID = 6435935558879109347L;
+    final List<FiringGroup> groupsAndTargets =
+        Aa.builder()
+            .aaUnits(firingUnits)
+            .aaTypes(aaTypes)
+            .hitPlayer(hitPlayer)
+            .attackableUnits(attackableUnits)
+            .defending(defending)
+            .gameData(bridge.getData())
+            .build()
+            .getFiringGroups();
 
-              @Override
-              public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
-                validTargets.removeAll(casualtiesSoFar);
-                if (!validTargets.isEmpty()) {
-                  dice =
-                      DiceRoll.rollAa(
-                          validTargets,
-                          firingGroup,
-                          allEnemyUnitsAliveOrWaitingToDie,
-                          allFriendlyUnitsAliveOrWaitingToDie,
-                          bridge,
-                          battleSite,
-                          defending);
-                  if (!headless) {
-                    SoundUtils.playFireBattleAa(firingPlayer, aaType, dice.getHits() > 0, bridge);
-                  }
+    for (final FiringGroup groupAndTarget : groupsAndTargets) {
+      final Collection<Unit> validTargets = groupAndTarget.getValidTargets();
+      final Collection<Unit> firingGroup = groupAndTarget.getFiringGroup();
+      final String aaType = groupAndTarget.getType();
+
+      final IExecutable rollDice =
+          new IExecutable() {
+            private static final long serialVersionUID = 6435935558879109347L;
+
+            @Override
+            public void execute(final ExecutionStack stack, final IDelegateBridge bridge) {
+              validTargets.removeAll(casualtiesSoFar);
+              if (!validTargets.isEmpty()) {
+                dice =
+                    DiceRoll.rollAa(
```

Comment:
Can this take FiringGroup to reduce number of params? (since validTargets, firingGroup, etc can just be taken from that param)

Reply:
I don't feel comfortable making a change to DiceRoll with this push.  DiceRoll#rollAa is also called from DiceRoll#rollSbrOrFlyOverAa so I'd have to change that as well.  And then I think having DiceRoll#rollAa using FiringGroup while DiceRoll#rollDice doesn't would feel weird.